File: src/main/java/in/dragonbra/javasteam/base/PacketClientGCMsg.java
package in.dragonbra.javasteam.base;

import in.dragonbra.javasteam.generated.MsgGCHdr;
import in.dragonbra.javasteam.types.JobID;

import java.io.ByteArrayInputStream;
import java.io.IOException;

/**
 * Represents a packet message with extended header information.
 */
public class PacketClientGCMsg implements IPacketGCMsg {

    private final int msgType;

    private final JobID targetJobID;

    private final JobID sourceJobID;

    private final byte[] payload;

    /**
     * Initializes a new instance of the {@link PacketClientGCMsg} class.
     *
     * @param eMsg The network message type for this packet message.
     * @param data The data.
     */
    public PacketClientGCMsg(int eMsg, byte[] data) {
        if (data == null) {
            throw new IllegalArgumentException("data is null");
        }

        msgType = eMsg;
        payload = data;

        MsgGCHdr gcHdr = new MsgGCHdr();

        // we need to pull out the job ids, so we deserialize the protobuf header
        try (ByteArrayInputStream bais = new ByteArrayInputStream(data)) {
            gcHdr.deserialize(bais);
        } catch (IOException ignored) {
        }

        targetJobID = new JobID(gcHdr.getTargetJobID());
        sourceJobID = new JobID(gcHdr.getSourceJobID());
    }

    @Override
    public boolean isProto() {
        return false;
    }

    @Override
    public int getMsgType() {
        return msgType;
    }

    @Override
    public JobID getTargetJobID() {
        return targetJobID;
    }

    @Override
    public JobID getSourceJobID() {
        return sourceJobID;
    }

    @Override
    public byte[] getData() {
        return payload;
    }
}

File: src/main/java/in/dragonbra/javasteam/base/gc/PacketClientGCMsg.kt
package `in`.dragonbra.javasteam.base.gc

import `in`.dragonbra.javasteam.generated.MsgGCHdr
import `in`.dragonbra.javasteam.types.JobID
import `in`.dragonbra.javasteam.util.stream.MemoryStream
import java.io.IOException

/**
 * Represents a packet message with extended header information.
 * @constructor Initializes a new instance of the [PacketClientGCMsg] class.
 * @param eMsg The network message type for this packet message.
 * @param data The data.
 */
class PacketClientGCMsg(private val eMsg: Int, data: ByteArray) : IPacketGCMsg {

    private val gcHdr: MsgGCHdr = MsgGCHdr()

    private val payload: ByteArray = data

    /**
     * Gets a value indicating whether this packet message is protobuf backed.
     * This type of message is never protobuf backed.
     */
    override val isProto: Boolean
        get() = false

    /**
     *  Gets the network message type of this packet message.
     */
    override val msgType: Int
        get() = eMsg

    /**
     * Gets the target job id for this packet message.
     */
    override val targetJobID: JobID
        get() = JobID(gcHdr.targetJobID)

    /**
     * Gets the source job id for this packet message.
     */
    override val sourceJobID: JobID
        get() = JobID(gcHdr.sourceJobID)

    /**
     * Gets the underlying data that represents this packet message.
     */
    override val data: ByteArray
        get() = payload

    init {
        // deserialize the gc header to get our hands on the job ids
        try {
            MemoryStream(data).use { ms ->
                gcHdr.deserialize(ms)
            }
        } catch (_: IOException) {
        }
    }
}
